cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -160,89 +150,62 @@ protected void doStartUp() {
    * channel and processes the TetheringControlResponses.
    */
   @Override
-  protected long runTask() {
-    List<PeerInfo> peers;
-    try {
-      peers = store.getPeers().stream()
-        // Ignore peers in REJECTED state.
-        .filter(p -> p.getTetheringStatus() != TetheringStatus.REJECTED)
-        .collect(Collectors.toList());
-    } catch (IOException e) {
-      LOG.warn("Failed to get peer information", e);
-      return connectionInterval;
-    }
-    Map<String, List<Notification>> peerToNotifications = getPeerProgramUpdates();
+  protected long runTask() throws IOException {
+    List<PeerInfo> peers = store.getPeers().stream()
+      // Ignore peers in REJECTED state.
+      .filter(p -> p.getTetheringStatus() != TetheringStatus.REJECTED)
+      .collect(Collectors.toList());
+
+    PeerProgramUpdates peerProgramUpdates = getPeerProgramUpdates(lastProgramUpdateMessageId);
     for (PeerInfo peer : peers) {
+      Preconditions.checkArgument(peer.getEndpoint() != null,
```

Comment:
would be good to add a comment explaining that this should never happen, as endpoint should only be null on the server side.

Reply:
done